test/e2e/rollback_byo_test.go

```diff
@@ -0,0 +1,174 @@
+	cancel := logstream.Start(t)
+	defer cancel()
+
+	clients := Setup(t)
+
+	serviceName := test.ObjectNameForTest(t)
+	byoNameOld := serviceName + "-byo-foo"
+	byoNameNew := serviceName + "-byo-foo-new"
+	names := test.ResourceNames{
+		Service: serviceName,
+		Image:   "helloworld",
+		BYOName: byoNameOld,
+	}
+
+	test.CleanupOnInterrupt(func() { test.TearDown(clients, names) })
+	defer test.TearDown(clients, names)
+
+	withTrafficSpecOld := WithInlineRouteSpec(v1alpha1.RouteSpec{
+		Traffic: []v1alpha1.TrafficTarget{
+			{
+				TrafficTarget: v1.TrafficTarget{
+					RevisionName: byoNameOld,
+					Percent:      ptr.Int64(100),
+				},
+			},
+		},
+	})
+	withTrafficSpecNew := WithInlineRouteSpec(v1alpha1.RouteSpec{
+		Traffic: []v1alpha1.TrafficTarget{
+			{
+				TrafficTarget: v1.TrafficTarget{
+					RevisionName: byoNameNew,
+					Percent:      ptr.Int64(100),
+				},
+			},
+		},
+	})
+
+	t.Logf("Creating a new Service with byo config name %q.", byoNameOld)
+	resources, err := v1a1test.CreateRunLatestServiceReady(t, clients, &names, withTrafficSpecOld)
+	if err != nil {
+		t.Fatalf("Failed to create initial Service: %v: %v", names.Service, err)
+	}
+	url := resources.Route.Status.URL.URL()
+	if _, err = pkgTest.WaitForEndpointState(
+		clients.KubeClient,
+		t.Logf,
+		url,
+		v1a1test.RetryingRouteInconsistency(pkgTest.MatchesAllOf(pkgTest.IsStatusOK, pkgTest.MatchesBody(test.HelloWorldText))),
+		"HelloWorldServesText",
+		test.ServingFlags.ResolvableDomain); err != nil {
+		t.Fatalf("The endpoint %s for Route %s didn't serve the expected text %q: %v", url, names.Route, test.HelloWorldText, err)
+	}
+
+	revisionName := resources.Revision.ObjectMeta.Name
+	if revisionName != byoNameOld {
+		t.Fatalf("Expect configuration name in revision label %q but got %q ", byoNameOld, revisionName)
+	}
+
+	// Update service to use a new byo name
+	t.Logf("Updating the Service to a new revision with a new byo name %q.", byoNameNew)
+	newSvc := resources.Service.DeepCopy()
+	so := WithInlineConfigSpec(*v1a1test.ConfigurationSpec(pkgTest.ImagePath(test.PizzaPlanet1), &byoNameNew))
+	so(newSvc)
+	withTrafficSpecNew(newSvc)
+	svc, err := v1a1test.PatchService(t, clients, resources.Service, newSvc)
+	resources.Service = svc
+	if err != nil {
+		t.Fatalf("Patch update for Service (new byo name %q) failed: %v", byoNameNew, err)
+	}
+
+	t.Log("Since the Service was updated a new Revision will be created and the Service will be updated")
+	newRevision, err := v1a1test.WaitForServiceLatestRevision(clients, names)
+	if err != nil {
+		t.Fatalf("Service %s was not updated with the Revision for new byo name %s: %v", names.Service, byoNameNew, err)
+	}
+	if newRevision != byoNameNew {
+		t.Fatalf("Expect configuration name in revision label %q but got %q ", byoNameNew, newRevision)
+	}
+
+	// Now, rollback to the first RevisionSpec
+	rollbackSvc := resources.Service.DeepCopy()
+	rollbackSvc.Spec = v1alpha1.ServiceSpec{
+		DeprecatedRelease: &v1alpha1.ReleaseType{
```

Comment:
When we rollback the Service's spec should be identical to when it was first created.

Reply:
The whole point of the test is that we should be able to rollback the *whole* spec (e.g. GitOps) and things should still work.  So the progression should be:

First:
```yaml
spec: 
  template:
    metadata:
      name: foo
    spec:
       containers:
       - image: foo
  traffic:
  - revisionName: foo
    percent: 100
```

Second:
```yaml
spec: 
  template:
    metadata:
      name: bar
    spec:
       containers:
       - image: bar
  traffic:
  - revisionName: bar
    percent: 100
```

Third:
```yaml
spec: 
  template:
    metadata:
      name: foo
    spec:
       containers:
       - image: foo
  traffic:
  - revisionName: foo
    percent: 100
```

However, what you have (ignoring release mode) is more like:
```yaml
spec: 
  template:
    metadata:
      name: bar
    spec:
       containers:
       - image: bar
  traffic:
  - revisionName: foo
    percent: 100
```

The whole point is that we *should* be able to achieve the third state by simply re-applying the initial state.